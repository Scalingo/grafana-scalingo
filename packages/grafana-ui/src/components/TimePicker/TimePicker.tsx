// Libraries
import React, { PureComponent, memo, FormEvent } from 'react';
import { css, cx } from 'emotion';

// Components
import { Tooltip } from '../Tooltip/Tooltip';
import { TimePickerContent } from './TimePickerContent/TimePickerContent';
import { ClickOutsideWrapper } from '../ClickOutsideWrapper/ClickOutsideWrapper';

// Utils & Services
import { stylesFactory } from '../../themes/stylesFactory';
import { withTheme, useTheme } from '../../themes/ThemeContext';

// Types
import { isDateTime, DateTime, rangeUtil, GrafanaTheme, TIME_FORMAT } from '@grafana/data';
import { TimeRange, TimeOption, TimeZone, dateMath } from '@grafana/data';
import { Themeable } from '../../types';

const quickOptions: TimeOption[] = [
  { from: 'now-5m', to: 'now', display: 'Last 5 minutes', section: 3 },
  { from: 'now-15m', to: 'now', display: 'Last 15 minutes', section: 3 },
  { from: 'now-30m', to: 'now', display: 'Last 30 minutes', section: 3 },
  { from: 'now-1h', to: 'now', display: 'Last 1 hour', section: 3 },
  { from: 'now-3h', to: 'now', display: 'Last 3 hours', section: 3 },
  { from: 'now-6h', to: 'now', display: 'Last 6 hours', section: 3 },
  { from: 'now-12h', to: 'now', display: 'Last 12 hours', section: 3 },
  { from: 'now-24h', to: 'now', display: 'Last 24 hours', section: 3 },
  { from: 'now-2d', to: 'now', display: 'Last 2 days', section: 3 },
  { from: 'now-7d', to: 'now', display: 'Last 7 days', section: 3 },
  { from: 'now-30d', to: 'now', display: 'Last 30 days', section: 3 },
  { from: 'now-90d', to: 'now', display: 'Last 90 days', section: 3 },
  { from: 'now-6M', to: 'now', display: 'Last 6 months', section: 3 },
  { from: 'now-1y', to: 'now', display: 'Last 1 year', section: 3 },
  { from: 'now-2y', to: 'now', display: 'Last 2 years', section: 3 },
  { from: 'now-5y', to: 'now', display: 'Last 5 years', section: 3 },
];

const otherOptions: TimeOption[] = [
  { from: 'now-1d/d', to: 'now-1d/d', display: 'Yesterday', section: 3 },
  { from: 'now-2d/d', to: 'now-2d/d', display: 'Day before yesterday', section: 3 },
  { from: 'now-7d/d', to: 'now-7d/d', display: 'This day last week', section: 3 },
  { from: 'now-1w/w', to: 'now-1w/w', display: 'Previous week', section: 3 },
  { from: 'now-1M/M', to: 'now-1M/M', display: 'Previous month', section: 3 },
  { from: 'now-1y/y', to: 'now-1y/y', display: 'Previous year', section: 3 },
  { from: 'now/d', to: 'now/d', display: 'Today', section: 3 },
  { from: 'now/d', to: 'now', display: 'Today so far', section: 3 },
  { from: 'now/w', to: 'now/w', display: 'This week', section: 3 },
  { from: 'now/w', to: 'now', display: 'This week so far', section: 3 },
  { from: 'now/M', to: 'now/M', display: 'This month', section: 3 },
  { from: 'now/M', to: 'now', display: 'This month so far', section: 3 },
  { from: 'now/y', to: 'now/y', display: 'This year', section: 3 },
  { from: 'now/y', to: 'now', display: 'This year so far', section: 3 },
];

const getStyles = stylesFactory((theme: GrafanaTheme) => {
  return {
    container: css`
      position: relative;
      display: flex;
      flex-flow: column nowrap;
    `,
    buttons: css`
      display: flex;
    `,
    caretIcon: css`
      margin-left: 3px;

      i {
        font-size: ${theme.typography.size.md};
      }
    `,
    syncedTimePicker: css`
      label: syncedTimePicker;
      border-color: ${theme.colors.orangeDark};
      background-image: none;
      background-color: transparent;
      color: ${theme.colors.orangeDark};
      &:focus,
      :hover {
        color: ${theme.colors.orangeDark};
        background-image: none;
        background-color: transparent;
      }
    `,
    noRightBorderStyle: css`
      label: noRightBorderStyle;
      border-right: 0;
    `,
  };
});

const getLabelStyles = stylesFactory((theme: GrafanaTheme) => {
  return {
    container: css`
      display: inline-block;
    `,
    utc: css`
      color: ${theme.colors.orange};
      font-size: 75%;
      padding: 3px;
      font-weight: ${theme.typography.weight.semibold};
    `,
  };
});

export interface Props extends Themeable {
  hideText?: boolean;
  value: TimeRange;
  timeZone?: TimeZone;
  timeSyncButton?: JSX.Element;
  isSynced?: boolean;
  onChange: (timeRange: TimeRange) => void;
  onMoveBackward: () => void;
  onMoveForward: () => void;
  onZoom: () => void;
  history?: TimeRange[];
}

export interface State {
  isOpen: boolean;
}

export class UnthemedTimePicker extends PureComponent<Props, State> {
  state: State = {
    isOpen: false,
  };

  onChange = (timeRange: TimeRange) => {
    this.props.onChange(timeRange);
    this.setState({ isOpen: false });
  };

  onOpen = (event: FormEvent<HTMLButtonElement>) => {
    const { isOpen } = this.state;
    event.stopPropagation();
    this.setState({ isOpen: !isOpen });
  };

  onClose = () => {
    this.setState({ isOpen: false });
  };

  render() {
    const {
      value,
      onMoveBackward,
      onMoveForward,
      onZoom,
      timeZone,
      timeSyncButton,
      isSynced,
      theme,
      history,
    } = this.props;

    const { isOpen } = this.state;
    const styles = getStyles(theme);
    const hasAbsolute = isDateTime(value.raw.from) || isDateTime(value.raw.to);
    const syncedTimePicker = timeSyncButton && isSynced;
    const timePickerIconClass = cx('fa fa-clock-o fa-fw', { ['icon-brand-gradient']: syncedTimePicker });
    const timePickerButtonClass = cx('btn navbar-button navbar-button--tight', {
      [`btn--radius-right-0 ${styles.noRightBorderStyle}`]: !!timeSyncButton,
      [`explore-active-button-glow ${styles.syncedTimePicker}`]: syncedTimePicker,
    });

    return (
      <div className={styles.container}>
        <div className={styles.buttons}>
          {hasAbsolute && (
            <button className="btn navbar-button navbar-button--tight" onClick={onMoveBackward}>
              <i className="fa fa-chevron-left" />
            </button>
          )}
          <div>
            <Tooltip content={<TimePickerTooltip timeRange={value} />} placement="bottom">
              <button aria-label="TimePicker Open Button" className={timePickerButtonClass} onClick={this.onOpen}>
                <i className={timePickerIconClass} />
                <TimePickerButtonLabel {...this.props} />
                <span className={styles.caretIcon}>
                  {isOpen ? <i className="fa fa-caret-up fa-fw" /> : <i className="fa fa-caret-down fa-fw" />}
                </span>
              </button>
            </Tooltip>
            {isOpen && (
              <ClickOutsideWrapper onClick={this.onClose}>
                <TimePickerContent
                  timeZone={timeZone}
                  value={value}
                  onChange={this.onChange}
                  otherOptions={otherOptions}
                  quickOptions={quickOptions}
                  history={history}
                />
              </ClickOutsideWrapper>
            )}
          </div>

          {timeSyncButton}

          {hasAbsolute && (
            <button className="btn navbar-button navbar-button--tight" onClick={onMoveForward}>
              <i className="fa fa-chevron-right" />
            </button>
          )}

          <Tooltip content={ZoomOutTooltip} placement="bottom">
            <button className="btn navbar-button navbar-button--zoom" onClick={onZoom}>
              <i className="fa fa-search-minus" />
            </button>
          </Tooltip>
        </div>
      </div>
    );
  }
}

const ZoomOutTooltip = () => (
  <>
    Time range zoom out <br /> CTRL+Z
  </>
);

const TimePickerTooltip = ({ timeRange }: { timeRange: TimeRange }) => (
  <>
    {timeRange.from.format(TIME_FORMAT)}
    <div className="text-center">to</div>
    {timeRange.to.format(TIME_FORMAT)}
  </>
);

const TimePickerButtonLabel = memo<Props>(props => {
  const theme = useTheme();
  const styles = getLabelStyles(theme);
  const isUTC = props.timeZone === 'utc';

  if (props.hideText) {
    return null;
  }

  return (
    <span className={styles.container}>
      <span>{formattedRange(props.value, isUTC)}</span>
      {isUTC && <span className={styles.utc}>UTC</span>}
    </span>
  );
});

const formattedRange = (value: TimeRange, isUTC: boolean) => {
  const adjustedTimeRange = {
    to: dateMath.isMathString(value.raw.to) ? value.raw.to : adjustedTime(value.to, isUTC),
    from: dateMath.isMathString(value.raw.from) ? value.raw.from : adjustedTime(value.from, isUTC),
  };
  return rangeUtil.describeTimeRange(adjustedTimeRange);
};

const adjustedTime = (time: DateTime, isUTC: boolean) => {
  if (isUTC) {
    return time.utc() || null;
  }
  return time.local() || null;
};

export const TimePicker = withTheme(UnthemedTimePicker);
