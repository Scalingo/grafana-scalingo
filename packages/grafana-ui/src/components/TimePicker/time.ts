import moment, { Moment } from 'moment';
import { TimeOption, TimeRange, TIME_FORMAT } from '../../types/time';
import { describeTimeRange } from '../../utils/rangeutil';
import * as dateMath from '../../utils/datemath';

export const mapTimeOptionToTimeRange = (
  timeOption: TimeOption,
  isTimezoneUtc: boolean,
  timezone?: dateMath.Timezone
): TimeRange => {
  const fromMoment = stringToMoment(timeOption.from, isTimezoneUtc, false, timezone);
  const toMoment = stringToMoment(timeOption.to, isTimezoneUtc, true, timezone);

  return { from: fromMoment, to: toMoment, raw: { from: timeOption.from, to: timeOption.to } };
};

export const stringToMoment = (
  value: string,
  isTimezoneUtc: boolean,
  roundUp?: boolean,
  timezone?: dateMath.Timezone
): Moment => {
  if (value.indexOf('now') !== -1) {
    if (!dateMath.isValid(value)) {
      return moment();
    }

    const parsed = dateMath.parse(value, roundUp, timezone);
    return parsed || moment();
  }

  if (isTimezoneUtc) {
    return moment.utc(value, TIME_FORMAT);
  }

  return moment(value, TIME_FORMAT);
};

export const mapTimeRangeToRangeString = (timeRange: TimeRange): string => {
  return describeTimeRange(timeRange.raw);
};

export const isValidTimeString = (text: string) => dateMath.isValid(text);
