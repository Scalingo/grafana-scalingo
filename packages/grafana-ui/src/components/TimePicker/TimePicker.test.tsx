import React from 'react';
import { shallow } from 'enzyme';
import { UnthemedTimePicker } from './TimePicker';
import { dateTime, TimeRange } from '@grafana/data';
import dark from './../../themes/dark';

const from = '2019-12-17T07:48:27.433Z';
const to = '2019-12-18T07:48:27.433Z';

const value: TimeRange = {
  from: dateTime(from),
  to: dateTime(to),
  raw: { from: dateTime(from), to: dateTime(to) },
};

describe('TimePicker', () => {
  it('renders buttons correctly', () => {
    const wrapper = shallow(
      <UnthemedTimePicker
        onChange={value => {}}
        value={value}
        onMoveBackward={() => {}}
        onMoveForward={() => {}}
        onZoom={() => {}}
        theme={dark}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders content correctly after beeing open', () => {
    const wrapper = shallow(
      <UnthemedTimePicker
        onChange={value => {}}
        value={value}
        onMoveBackward={() => {}}
        onMoveForward={() => {}}
        onZoom={() => {}}
        theme={dark}
      />
    );

    wrapper.find('button[aria-label="TimePicker Open Button"]').simulate('click', new Event('click'));
    expect(wrapper).toMatchSnapshot();
  });
});
