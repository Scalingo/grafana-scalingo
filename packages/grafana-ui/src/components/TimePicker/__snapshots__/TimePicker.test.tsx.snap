// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker renders buttons correctly 1`] = `
<div
  className="css-16ba5ut"
>
  <div
    className="css-vyoujf"
  >
    <button
      className="btn navbar-button navbar-button--tight"
      onClick={[Function]}
    >
      <i
        className="fa fa-chevron-left"
      />
    </button>
    <div>
      <Component
        content={
          <TimePickerTooltip
            timeRange={
              Object {
                "from": "2019-12-17T07:48:27.433Z",
                "raw": Object {
                  "from": "2019-12-17T07:48:27.433Z",
                  "to": "2019-12-18T07:48:27.433Z",
                },
                "to": "2019-12-18T07:48:27.433Z",
              }
            }
          />
        }
        placement="bottom"
      >
        <button
          aria-label="TimePicker Open Button"
          className="btn navbar-button navbar-button--tight"
          onClick={[Function]}
        >
          <i
            className="fa fa-clock-o fa-fw"
          />
          <Component
            onChange={[Function]}
            onMoveBackward={[Function]}
            onMoveForward={[Function]}
            onZoom={[Function]}
            theme={
              Object {
                "background": Object {
                  "dropdown": "#1f1f20",
                  "pageHeader": "linear-gradient(90deg, #292a2d, #000000)",
                  "scrollbar": "#343436",
                  "scrollbar2": "#343436",
                },
                "border": Object {
                  "radius": Object {
                    "lg": "5px",
                    "md": "3px",
                    "sm": "2px",
                  },
                  "width": Object {
                    "sm": "1px",
                  },
                },
                "breakpoints": Object {
                  "lg": "992px",
                  "md": "769px",
                  "sm": "544px",
                  "xl": "1200px",
                  "xs": "0",
                },
                "colors": Object {
                  "black": "#000000",
                  "blue": "#33b5e5",
                  "blue77": "#1f60c4",
                  "blue85": "#3274d9",
                  "blue95": "#5794f2",
                  "blueBase": "#3274d9",
                  "blueFaint": "#041126",
                  "blueLight": "#5794f2",
                  "blueShade": "#1f60c4",
                  "body": "#d8d9da",
                  "bodyBg": "#161719",
                  "brandDanger": "#e02f44",
                  "brandPrimary": "#eb7b18",
                  "brandSuccess": "#299c46",
                  "brandWarning": "#eb7b18",
                  "critical": "#e02f44",
                  "dark1": "#141414",
                  "dark10": "#424345",
                  "dark2": "#161719",
                  "dark3": "#1f1f20",
                  "dark4": "#212124",
                  "dark5": "#222426",
                  "dark6": "#262628",
                  "dark7": "#292a2d",
                  "dark8": "#2f2f32",
                  "dark9": "#343436",
                  "formCheckboxBg": "#222426",
                  "formCheckboxBgChecked": "#5794f2",
                  "formCheckboxBgCheckedHover": "#3274d9",
                  "formCheckboxCheckmark": "#343b40",
                  "formDescription": "#9fa7b3",
                  "formFocusOutline": "#1f60c4",
                  "formInputBg": "#202226",
                  "formInputBgDisabled": "#141619",
                  "formInputBorder": "#343b40",
                  "formInputBorderActive": "#5794f2",
                  "formInputBorderHover": "#464c54",
                  "formInputBorderInvalid": "#e02f44",
                  "formInputDisabledText": "#9fa7b3",
                  "formInputText": "#c7d0d9",
                  "formInputTextStrong": "#c7d0d9",
                  "formInputTextWhite": "#ffffff",
                  "formLabel": "#9fa7b3",
                  "formLegend": "#c7d0d9",
                  "formSwitchBg": "#343b40",
                  "formSwitchBgActive": "#5794f2",
                  "formSwitchBgActiveHover": "#3274d9",
                  "formSwitchBgDisabled": "#343b40",
                  "formSwitchBgHover": "#464c54",
                  "formSwitchDot": "#202226",
                  "formValidationMessageBg": "#e02f44",
                  "formValidationMessageText": "#ffffff",
                  "gray05": "#0b0c0e",
                  "gray1": "#555555",
                  "gray10": "#141619",
                  "gray15": "#202226",
                  "gray2": "#8e8e8e",
                  "gray25": "#343b40",
                  "gray3": "#b3b3b3",
                  "gray33": "#464c54",
                  "gray4": "#d8d9da",
                  "gray5": "#ececec",
                  "gray6": "#f4f5f8",
                  "gray7": "#fbfbfb",
                  "gray70": "#9fa7b3",
                  "gray85": "#c7d0d9",
                  "gray95": "#e9edf2",
                  "gray98": "#f7f8fa",
                  "grayBlue": "#212327",
                  "greenBase": "#299c46",
                  "greenShade": "#23843b",
                  "headingColor": "#d8d9da",
                  "inputBlack": "#09090b",
                  "link": "#d8d9da",
                  "linkDisabled": "#8e8e8e",
                  "linkExternal": "#33b5e5",
                  "linkHover": "#ffffff",
                  "online": "#299c46",
                  "orange": "#eb7b18",
                  "orangeDark": "#ff780a",
                  "pageBg": "#161719",
                  "pageHeaderBorder": "#343436",
                  "panelBg": "#212124",
                  "purple": "#9933cc",
                  "queryGreen": "#74e680",
                  "queryKeyword": "#66d9ef",
                  "queryOrange": "#eb7b18",
                  "queryPurple": "#fe85fc",
                  "queryRed": "#e02f44",
                  "red": "#d44a3a",
                  "red88": "#e02f44",
                  "redBase": "#e02f44",
                  "redShade": "#c4162a",
                  "text": "#d8d9da",
                  "textEmphasis": "#ececec",
                  "textFaint": "#222426",
                  "textStrong": "#ffffff",
                  "textWeak": "#8e8e8e",
                  "variable": "#32d1df",
                  "warn": "#f79520",
                  "white": "#ffffff",
                  "yellow": "#ecbb13",
                },
                "height": Object {
                  "lg": "48px",
                  "md": "32px",
                  "sm": "24px",
                },
                "isDark": true,
                "isLight": false,
                "name": "Grafana Dark",
                "panelHeaderHeight": 28,
                "panelPadding": 8,
                "shadow": Object {
                  "pageHeader": "inset 0px -4px 14px #1f1f20",
                },
                "spacing": Object {
                  "d": "14px",
                  "formButtonHeight": 32,
                  "formFieldsetMargin": "16px",
                  "formInputAffixPaddingHorizontal": "4px",
                  "formInputHeight": "32px",
                  "formInputMargin": "16px",
                  "formInputPaddingHorizontal": "8px",
                  "formLabelMargin": "0 0 4px 0",
                  "formLabelPadding": "0 0 0 2px",
                  "formLegendMargin": "0 0 16px 0",
                  "formMargin": "32px",
                  "formSpacingBase": 8,
                  "formValidationMessageMargin": "4px 0 0 0",
                  "formValidationMessagePadding": "4px 8px",
                  "gutter": "30px",
                  "insetSquishMd": "4px 8px",
                  "lg": "24px",
                  "md": "16px",
                  "sm": "8px",
                  "xl": "32px",
                  "xs": "4px",
                  "xxs": "2px",
                },
                "type": "dark",
                "typography": Object {
                  "fontFamily": Object {
                    "monospace": "Menlo, Monaco, Consolas, 'Courier New', monospace",
                    "sansSerif": "'Roboto', 'Helvetica Neue', Arial, sans-serif",
                  },
                  "heading": Object {
                    "h1": "28px",
                    "h2": "24px",
                    "h3": "21px",
                    "h4": "18px",
                    "h5": "16px",
                    "h6": "14px",
                  },
                  "lineHeight": Object {
                    "lg": 2,
                    "md": 1.5,
                    "sm": 1.1,
                    "xs": 1,
                  },
                  "link": Object {
                    "decoration": "none",
                    "hoverDecoration": "none",
                  },
                  "size": Object {
                    "base": "14px",
                    "lg": "18px",
                    "md": "14px",
                    "sm": "12px",
                    "xs": "10px",
                  },
                  "weight": Object {
                    "bold": 600,
                    "light": 300,
                    "regular": 400,
                    "semibold": 500,
                  },
                },
                "zIndex": Object {
                  "dropdown": "1000",
                  "modal": "1050",
                  "modalBackdrop": "1040",
                  "navbarFixed": "1020",
                  "sidemenu": "1025",
                  "tooltip": "1030",
                  "typeahead": "1060",
                },
              }
            }
            value={
              Object {
                "from": "2019-12-17T07:48:27.433Z",
                "raw": Object {
                  "from": "2019-12-17T07:48:27.433Z",
                  "to": "2019-12-18T07:48:27.433Z",
                },
                "to": "2019-12-18T07:48:27.433Z",
              }
            }
          />
          <span
            className="css-6x26ye"
          >
            <i
              className="fa fa-caret-down fa-fw"
            />
          </span>
        </button>
      </Component>
    </div>
    <button
      className="btn navbar-button navbar-button--tight"
      onClick={[Function]}
    >
      <i
        className="fa fa-chevron-right"
      />
    </button>
    <Component
      content={[Function]}
      placement="bottom"
    >
      <button
        className="btn navbar-button navbar-button--zoom"
        onClick={[Function]}
      >
        <i
          className="fa fa-search-minus"
        />
      </button>
    </Component>
  </div>
</div>
`;

exports[`TimePicker renders content correctly after beeing open 1`] = `
<div
  className="css-16ba5ut"
>
  <div
    className="css-vyoujf"
  >
    <button
      className="btn navbar-button navbar-button--tight"
      onClick={[Function]}
    >
      <i
        className="fa fa-chevron-left"
      />
    </button>
    <div>
      <Component
        content={
          <TimePickerTooltip
            timeRange={
              Object {
                "from": "2019-12-17T07:48:27.433Z",
                "raw": Object {
                  "from": "2019-12-17T07:48:27.433Z",
                  "to": "2019-12-18T07:48:27.433Z",
                },
                "to": "2019-12-18T07:48:27.433Z",
              }
            }
          />
        }
        placement="bottom"
      >
        <button
          aria-label="TimePicker Open Button"
          className="btn navbar-button navbar-button--tight"
          onClick={[Function]}
        >
          <i
            className="fa fa-clock-o fa-fw"
          />
          <Component
            onChange={[Function]}
            onMoveBackward={[Function]}
            onMoveForward={[Function]}
            onZoom={[Function]}
            theme={
              Object {
                "background": Object {
                  "dropdown": "#1f1f20",
                  "pageHeader": "linear-gradient(90deg, #292a2d, #000000)",
                  "scrollbar": "#343436",
                  "scrollbar2": "#343436",
                },
                "border": Object {
                  "radius": Object {
                    "lg": "5px",
                    "md": "3px",
                    "sm": "2px",
                  },
                  "width": Object {
                    "sm": "1px",
                  },
                },
                "breakpoints": Object {
                  "lg": "992px",
                  "md": "769px",
                  "sm": "544px",
                  "xl": "1200px",
                  "xs": "0",
                },
                "colors": Object {
                  "black": "#000000",
                  "blue": "#33b5e5",
                  "blue77": "#1f60c4",
                  "blue85": "#3274d9",
                  "blue95": "#5794f2",
                  "blueBase": "#3274d9",
                  "blueFaint": "#041126",
                  "blueLight": "#5794f2",
                  "blueShade": "#1f60c4",
                  "body": "#d8d9da",
                  "bodyBg": "#161719",
                  "brandDanger": "#e02f44",
                  "brandPrimary": "#eb7b18",
                  "brandSuccess": "#299c46",
                  "brandWarning": "#eb7b18",
                  "critical": "#e02f44",
                  "dark1": "#141414",
                  "dark10": "#424345",
                  "dark2": "#161719",
                  "dark3": "#1f1f20",
                  "dark4": "#212124",
                  "dark5": "#222426",
                  "dark6": "#262628",
                  "dark7": "#292a2d",
                  "dark8": "#2f2f32",
                  "dark9": "#343436",
                  "formCheckboxBg": "#222426",
                  "formCheckboxBgChecked": "#5794f2",
                  "formCheckboxBgCheckedHover": "#3274d9",
                  "formCheckboxCheckmark": "#343b40",
                  "formDescription": "#9fa7b3",
                  "formFocusOutline": "#1f60c4",
                  "formInputBg": "#202226",
                  "formInputBgDisabled": "#141619",
                  "formInputBorder": "#343b40",
                  "formInputBorderActive": "#5794f2",
                  "formInputBorderHover": "#464c54",
                  "formInputBorderInvalid": "#e02f44",
                  "formInputDisabledText": "#9fa7b3",
                  "formInputText": "#c7d0d9",
                  "formInputTextStrong": "#c7d0d9",
                  "formInputTextWhite": "#ffffff",
                  "formLabel": "#9fa7b3",
                  "formLegend": "#c7d0d9",
                  "formSwitchBg": "#343b40",
                  "formSwitchBgActive": "#5794f2",
                  "formSwitchBgActiveHover": "#3274d9",
                  "formSwitchBgDisabled": "#343b40",
                  "formSwitchBgHover": "#464c54",
                  "formSwitchDot": "#202226",
                  "formValidationMessageBg": "#e02f44",
                  "formValidationMessageText": "#ffffff",
                  "gray05": "#0b0c0e",
                  "gray1": "#555555",
                  "gray10": "#141619",
                  "gray15": "#202226",
                  "gray2": "#8e8e8e",
                  "gray25": "#343b40",
                  "gray3": "#b3b3b3",
                  "gray33": "#464c54",
                  "gray4": "#d8d9da",
                  "gray5": "#ececec",
                  "gray6": "#f4f5f8",
                  "gray7": "#fbfbfb",
                  "gray70": "#9fa7b3",
                  "gray85": "#c7d0d9",
                  "gray95": "#e9edf2",
                  "gray98": "#f7f8fa",
                  "grayBlue": "#212327",
                  "greenBase": "#299c46",
                  "greenShade": "#23843b",
                  "headingColor": "#d8d9da",
                  "inputBlack": "#09090b",
                  "link": "#d8d9da",
                  "linkDisabled": "#8e8e8e",
                  "linkExternal": "#33b5e5",
                  "linkHover": "#ffffff",
                  "online": "#299c46",
                  "orange": "#eb7b18",
                  "orangeDark": "#ff780a",
                  "pageBg": "#161719",
                  "pageHeaderBorder": "#343436",
                  "panelBg": "#212124",
                  "purple": "#9933cc",
                  "queryGreen": "#74e680",
                  "queryKeyword": "#66d9ef",
                  "queryOrange": "#eb7b18",
                  "queryPurple": "#fe85fc",
                  "queryRed": "#e02f44",
                  "red": "#d44a3a",
                  "red88": "#e02f44",
                  "redBase": "#e02f44",
                  "redShade": "#c4162a",
                  "text": "#d8d9da",
                  "textEmphasis": "#ececec",
                  "textFaint": "#222426",
                  "textStrong": "#ffffff",
                  "textWeak": "#8e8e8e",
                  "variable": "#32d1df",
                  "warn": "#f79520",
                  "white": "#ffffff",
                  "yellow": "#ecbb13",
                },
                "height": Object {
                  "lg": "48px",
                  "md": "32px",
                  "sm": "24px",
                },
                "isDark": true,
                "isLight": false,
                "name": "Grafana Dark",
                "panelHeaderHeight": 28,
                "panelPadding": 8,
                "shadow": Object {
                  "pageHeader": "inset 0px -4px 14px #1f1f20",
                },
                "spacing": Object {
                  "d": "14px",
                  "formButtonHeight": 32,
                  "formFieldsetMargin": "16px",
                  "formInputAffixPaddingHorizontal": "4px",
                  "formInputHeight": "32px",
                  "formInputMargin": "16px",
                  "formInputPaddingHorizontal": "8px",
                  "formLabelMargin": "0 0 4px 0",
                  "formLabelPadding": "0 0 0 2px",
                  "formLegendMargin": "0 0 16px 0",
                  "formMargin": "32px",
                  "formSpacingBase": 8,
                  "formValidationMessageMargin": "4px 0 0 0",
                  "formValidationMessagePadding": "4px 8px",
                  "gutter": "30px",
                  "insetSquishMd": "4px 8px",
                  "lg": "24px",
                  "md": "16px",
                  "sm": "8px",
                  "xl": "32px",
                  "xs": "4px",
                  "xxs": "2px",
                },
                "type": "dark",
                "typography": Object {
                  "fontFamily": Object {
                    "monospace": "Menlo, Monaco, Consolas, 'Courier New', monospace",
                    "sansSerif": "'Roboto', 'Helvetica Neue', Arial, sans-serif",
                  },
                  "heading": Object {
                    "h1": "28px",
                    "h2": "24px",
                    "h3": "21px",
                    "h4": "18px",
                    "h5": "16px",
                    "h6": "14px",
                  },
                  "lineHeight": Object {
                    "lg": 2,
                    "md": 1.5,
                    "sm": 1.1,
                    "xs": 1,
                  },
                  "link": Object {
                    "decoration": "none",
                    "hoverDecoration": "none",
                  },
                  "size": Object {
                    "base": "14px",
                    "lg": "18px",
                    "md": "14px",
                    "sm": "12px",
                    "xs": "10px",
                  },
                  "weight": Object {
                    "bold": 600,
                    "light": 300,
                    "regular": 400,
                    "semibold": 500,
                  },
                },
                "zIndex": Object {
                  "dropdown": "1000",
                  "modal": "1050",
                  "modalBackdrop": "1040",
                  "navbarFixed": "1020",
                  "sidemenu": "1025",
                  "tooltip": "1030",
                  "typeahead": "1060",
                },
              }
            }
            value={
              Object {
                "from": "2019-12-17T07:48:27.433Z",
                "raw": Object {
                  "from": "2019-12-17T07:48:27.433Z",
                  "to": "2019-12-18T07:48:27.433Z",
                },
                "to": "2019-12-18T07:48:27.433Z",
              }
            }
          />
          <span
            className="css-6x26ye"
          >
            <i
              className="fa fa-caret-up fa-fw"
            />
          </span>
        </button>
      </Component>
      <ClickOutsideWrapper
        onClick={[Function]}
      >
        <Component
          onChange={[Function]}
          otherOptions={
            Array [
              Object {
                "display": "Yesterday",
                "from": "now-1d/d",
                "section": 3,
                "to": "now-1d/d",
              },
              Object {
                "display": "Day before yesterday",
                "from": "now-2d/d",
                "section": 3,
                "to": "now-2d/d",
              },
              Object {
                "display": "This day last week",
                "from": "now-7d/d",
                "section": 3,
                "to": "now-7d/d",
              },
              Object {
                "display": "Previous week",
                "from": "now-1w/w",
                "section": 3,
                "to": "now-1w/w",
              },
              Object {
                "display": "Previous month",
                "from": "now-1M/M",
                "section": 3,
                "to": "now-1M/M",
              },
              Object {
                "display": "Previous year",
                "from": "now-1y/y",
                "section": 3,
                "to": "now-1y/y",
              },
              Object {
                "display": "Today",
                "from": "now/d",
                "section": 3,
                "to": "now/d",
              },
              Object {
                "display": "Today so far",
                "from": "now/d",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "This week",
                "from": "now/w",
                "section": 3,
                "to": "now/w",
              },
              Object {
                "display": "This week so far",
                "from": "now/w",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "This month",
                "from": "now/M",
                "section": 3,
                "to": "now/M",
              },
              Object {
                "display": "This month so far",
                "from": "now/M",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "This year",
                "from": "now/y",
                "section": 3,
                "to": "now/y",
              },
              Object {
                "display": "This year so far",
                "from": "now/y",
                "section": 3,
                "to": "now",
              },
            ]
          }
          quickOptions={
            Array [
              Object {
                "display": "Last 5 minutes",
                "from": "now-5m",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 15 minutes",
                "from": "now-15m",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 30 minutes",
                "from": "now-30m",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 1 hour",
                "from": "now-1h",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 3 hours",
                "from": "now-3h",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 6 hours",
                "from": "now-6h",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 12 hours",
                "from": "now-12h",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 24 hours",
                "from": "now-24h",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 2 days",
                "from": "now-2d",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 7 days",
                "from": "now-7d",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 30 days",
                "from": "now-30d",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 90 days",
                "from": "now-90d",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 6 months",
                "from": "now-6M",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 1 year",
                "from": "now-1y",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 2 years",
                "from": "now-2y",
                "section": 3,
                "to": "now",
              },
              Object {
                "display": "Last 5 years",
                "from": "now-5y",
                "section": 3,
                "to": "now",
              },
            ]
          }
          value={
            Object {
              "from": "2019-12-17T07:48:27.433Z",
              "raw": Object {
                "from": "2019-12-17T07:48:27.433Z",
                "to": "2019-12-18T07:48:27.433Z",
              },
              "to": "2019-12-18T07:48:27.433Z",
            }
          }
        />
      </ClickOutsideWrapper>
    </div>
    <button
      className="btn navbar-button navbar-button--tight"
      onClick={[Function]}
    >
      <i
        className="fa fa-chevron-right"
      />
    </button>
    <Component
      content={[Function]}
      placement="bottom"
    >
      <button
        className="btn navbar-button navbar-button--zoom"
        onClick={[Function]}
      >
        <i
          className="fa fa-search-minus"
        />
      </button>
    </Component>
  </div>
</div>
`;
